packages/core/core/generate-text/stream-text.ts

```diff
@@ -0,0 +1,138 @@
+} from '../util/async-iterable-stream';
+import { retryWithExponentialBackoff } from '../util/retry-with-exponential-backoff';
+import { runToolsTransformation } from './run-tools-transformation';
+import { ToToolCall } from './tool-call';
+import { ToToolResult } from './tool-result';
+
+/**
+ * Stream text generated by a language model.
+ */
+export async function experimental_streamText<
+  TOOLS extends Record<string, ExperimentalTool>,
+>({
+  model,
+  tools,
+  system,
+  prompt,
+  messages,
+  maxRetries,
+  abortSignal,
+  ...settings
+}: CallSettings &
+  Prompt & {
+    model: LanguageModelV1;
+    tools?: TOOLS;
+  }): Promise<StreamTextResult<TOOLS>> {
+  const retry = retryWithExponentialBackoff({ maxRetries });
+  const { stream, warnings } = await retry(() =>
+    model.doStream({
+      mode: {
+        type: 'regular',
+        tools:
+          tools == null
+            ? undefined
+            : Object.entries(tools).map(([name, tool]) => ({
+                type: 'function',
+                name,
+                description: tool.description,
+                parameters: zodToJsonSchema(tool.parameters),
+              })),
+      },
+      ...prepareCallSettings(settings),
+      inputFormat: getInputFormat({ prompt, messages }),
+      prompt: convertToLanguageModelPrompt({
+        system,
+        prompt,
+        messages,
+      }),
+      abortSignal,
+    }),
+  );
+
+  const toolStream = runToolsTransformation({
+    tools,
+    generatorStream: stream,
+  });
+
+  return new StreamTextResult(toolStream);
+}
+
+export type TextStreamPart<TOOLS extends Record<string, ExperimentalTool>> =
+  | {
+      type: 'text-delta';
+      textDelta: string;
+    }
+  | ({
+      type: 'tool-call';
+    } & ToToolCall<TOOLS>)
+  | {
+      type: 'error';
+      error: unknown;
+    }
+  | ({
+      type: 'tool-result';
+    } & ToToolResult<TOOLS>);
+
+export class StreamTextResult<TOOLS extends Record<string, ExperimentalTool>> {
+  private readonly originalStream: ReadableStream<TextStreamPart<TOOLS>>;
+
+  constructor(stream: ReadableStream<TextStreamPart<TOOLS>>) {
+    this.originalStream = stream;
+  }
+
+  get textStream(): AsyncIterableStream<string> {
+    return createAsyncIterableStream(this.originalStream, {
+      transform(chunk, controller) {
+        if (chunk.type === 'text-delta') {
+          // do not stream empty text deltas:
+          if (chunk.textDelta.length > 0) {
+            controller.enqueue(chunk.textDelta);
+          }
+        } else if (chunk.type === 'error') {
+          throw chunk.error;
+        }
+      },
+    });
+  }
+
+  get fullStream(): AsyncIterableStream<TextStreamPart<TOOLS>> {
+    return createAsyncIterableStream(this.originalStream, {
+      transform(chunk, controller) {
+        if (chunk.type === 'text-delta') {
+          // do not stream empty text deltas:
+          if (chunk.textDelta.length > 0) {
+            controller.enqueue(chunk);
+          }
```

Comment:
But it still looks like that the initial `textDelta` is always empty from the tests?

Reply:
It depends on the provider. Providers might have empty chunks, but we don't forward them.